File: 백준/Gold/7576. 토마토/토마토.py
import sys
input = sys.stdin.readline
sys.setrecursionlimit(10**5)

from collections import deque 

n, m = map(int, input().split())

graph = []

for _ in range(m):
  graph.append(list(map(int, input().split())))


dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1]
# print(graph)
  
def bfs(temp):
  queue = deque(temp)

  while queue:
    # print(queue)
    x, y = queue.popleft()
    for i in range(4):
      nx = x + dx[i]
      ny = y + dy[i]

      if nx < 0 or ny < 0 or nx>=m or ny>=n:
        # print("칸넘")
        continue
      if graph[nx][ny]==-1:
        continue

      if graph[nx][ny]==0:
        graph[nx][ny] = graph[x][y]+1
        queue.append((nx, ny))


temp=[]

for i in range(len(graph)):
  for j in range(len(graph[0])):
    if graph[i][j]==1:
      temp.append((i, j))

# print(temp)
# if len(temp)>1:
#   bfs(i[0], i[1])


bfs(temp)

answer = 0
for i in graph:
  if 0 in i:
    answer=0
    break
  for j in i:
    if j>answer:
      answer=j
print(answer-1)

# 1이 여러개인 경우를 못 풀었음
# 6 4
# 1 -1 0 0 0 0
# 0 -1 0 0 0 0
# 0 0 0 0 -1 0
# 0 0 0 0 -1 1

File: 백준/Gold/7576. 토마토/토마토.java
//BFS
import java.io.*;
import java.util.*;

class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        String[]temp = br.readLine().split(" ");
        int n = Integer.parseInt(temp[1]);
        int m = Integer.parseInt(temp[0]);

        int[][] map = new int[n][m];

        for (int i = 0; i < n; i++) {
            String[] temp2 = br.readLine().split(" ");
            for (int j = 0; j < m; j++) {
                map[i][j] = Integer.parseInt(temp2[j]);
            }
        }
//        System.out.println(Arrays.deepToString(map));

        Deque<int[]> q = new ArrayDeque<>();
        for(int i = 0; i < n; i++){
            for(int j = 0; j < m; j++){
                if(map[i][j] == 1){
                    q.offer(new int[]{i, j, 0});
                }
            }
        }

        int[] dx = {0, 0, -1, 1};
        int[] dy = {1, -1, 0, 0};
        int count = 0;

        while (!q.isEmpty()){
            int[] cur = q.poll();
            int x = cur[0];
            int y = cur[1];
            count=cur[2];

            for(int i = 0; i < 4; i++){
                int nx = x + dx[i];
                int ny = y + dy[i];
                if(nx>=0 && nx<n && ny>=0 && ny<m && map[nx][ny] == 0){
                    q.offer(new int[]{nx, ny, count+1});
                    map[nx][ny] = 1;
                }
            }
        }
        for(int[] M: map){
            for(int MM: M){
                if(MM==0){
                    System.out.println(-1);
                    System.exit(0);
                }
            }
        }
        bw.write(count+"\n");
        bw.flush();
        bw.close();
    }
}